### ISSUE TITLE
reviewing PDB tests

### ISSUE THREAD

**Issue by rob-miller** [biopython/biopython#3230]
As per discussion in #3220, I stupidly volunteered to look at the structure files used by the test_PDB_*.py files.  I've generated the attached spreadsheet which attempts to list the files and what they are used for.  

It is just clean-up work so only enhancement for maintainers, but overall my feeling remains that we could slim down to fewer test files - perhaps more judiciously selected against the current database to find some small, valid and particularly ugly/nasty/horrible disordered mess to actively trigger disorder handling by default. This might be another artificially constructed bespoke file.  The improvement would be to further rationalize the test_PDB set as @JoaoRodrigues started with #3220 and probably speed up and clean up, but also to provide substrate for future code additions to test on by default.

Also going through the tests I noticed that 1AS5.cif seems to be only used to check that mkdssp reads nonstandard residues correctly, but mkdssp is an external program (and 1as5.cif has been updated to label them correctly anyway).  Also I haven't worked with MSMS or the ResidueDepth code, but looks like 2XHE was only used for that test (now commented out) and the internal_coords test as a multi-chain sample with missing residues.  I wonder if the three other structures run through MSMS in this test actually trigger any different pathways in Biopython code?

[biopython_pdb_tests.xlsx](https://github.com/biopython/biopython/files/5112101/biopython_pdb_tests.xlsx)

**Comment by JoaoRodrigues** [comment]
I downloaded all XML files as well as an 'orthogonal' dataset that is trivial to parse. I'll write code now to compare the Biopython parsed structures to those, namely number of chains, residues, atoms, etc. Just to make sure we are parsing everything as it should be.

**Comment by JoaoRodrigues** [comment]
@rob-miller now that I have the benchmark running, I'll go over your excel file. Great job at noting down all the tests and what they need the files for. We can almost surely craft a handful files that cover all these cases and add in one or two 'naturally' weird ones to boot!

**Comment by rob-miller** [comment]
Cool.  I attach some (large-ish) log files from my pdb and cif test code that I haven't been able to get through yet.  
My code is at https://github.com/rob-miller/rtm-biopython-scripts but the key thing is that the columns indicate complication in the structure:

`file_no format idcode chain_id chain_count residues disordered_residues atoms disordered_atoms resolution missing_residues non-sequential_residues`

if a non-sequential residue is detected (residue number goes down instead of up, common for e.g. insertion of lysozyme into a GPCR), the the residue info is printed and this comes **_before_** the report line on the chain.  

I found it interesting to see after we both finished that your testing was the round-trip check (great idea!), while I focused on finding disordered / complicated examples.  Might just go back to that example @jgreener64 mentioned (1EN2) mentioned though. 
 
[pdb.log.gz](https://github.com/biopython/biopython/files/5140137/pdb.log.gz)
[cif.log.gz](https://github.com/biopython/biopython/files/5140134/cif.log.gz)

**Comment by JoaoRodrigues** [comment]
Thanks! Awesome work, it's great we had different ideas to check, will make for a nice cross-check!

My mmCIF dataset is still running (much slower parsing than PDBs) - I'll update when it's done! For now, here's the PDB result:

```
            n_models       n_chains       n_resids       n_atoms  \
count  166670.000000  166670.000000  166670.000000  1.666700e+05   
mean        2.290676       4.668147    1098.904632  8.160455e+03   
std         5.482005       8.744392    1530.454436  1.458483e+04   
min         1.000000       1.000000       2.000000  3.100000e+01   
25%         1.000000       1.000000     417.000000  2.419000e+03   
50%         1.000000       2.000000     702.000000  4.235000e+03   
75%         1.000000       4.000000    1285.000000  8.295000e+03   
max       640.000000     920.000000  178320.000000  1.290100e+06   

          read_time     write_time  
count  166670.00000  166670.000000  
mean        0.16762       0.049608  
std         0.29104       0.086674  
min         0.00200       0.000000  
25%         0.04100       0.014000  
50%         0.07500       0.025000  
75%         0.18700       0.052000  
max        21.59100       7.276000
```

Also, I checked: there is a structure with 640 models. Wow.

**Comment by JoaoRodrigues** [comment]
Some updates:
- The entire PDB parses just fine, 100% success!
- Element assignment is correct for nearly all entries in the PDB (see #3250)
- The mmCIF dataset is underway (2/3 done) and it seems to be going well. There are a few instances of duplicated residues/atoms but I think actually these are issues with the files themselves. @jgreener64 would you mind sharing your PDB contact so I can email them about it? 
- Two things that came out of the mmCIF testing:
  - the mmCIF parser does not have a PERMISSIVE mode. Is there a particular reason for this?
  - the line numbers reported by StructureBuilder when parsing through MMFCIFParser are bogus since we're not reading the file per se, but the mmcif dictionary. We should consider either silencing these or setting them somehow from the cif parser to give more accurate pointers.

**Comment by peterjc** [comment]
I would guess the MMCIF parser didn't need a permissive parser since these files will be far more tightly curated than PDB format, where there was more variation in the early days from submissions and third party tool's output?

**Comment by JoaoRodrigues** [comment]
Maybe indeed, but we do see the same "issues" in the files. I'd be in favor of adding a permissive mode to that parser too, if only for consistency between all the inputs.

As for results of the benchmark. Both PDB and mmCIF parsers go through the entire PDB (bar those PERMISSIVE CIF files) and the resulting `Structure` objects match. This is very nice!! I now need to compare orthogonally to PDBML to make sure we parsed the right stuff. Then I'll compare with @rob-miller's results and also produce some timings/plots.

**Comment by jgreener64** [comment]
> would you mind sharing your PDB contact so I can email them about it?

I just use the "Question or Comment" form at https://www.rcsb.org/pages/contactus, they seem to be pretty responsive.

**Comment by JoaoRodrigues** [comment]
Thanks @jgreener64 !

I keep finding new issues with files: e.g. NMR models with different numbers of atoms. Maybe we can market Bio.PDB as a PDB validator :)

**Comment by rob-miller** [comment]
Seems like the few mmCIF problems I looked at that had PDB equivalents had
the same issues, just mmCIF did not have the permissive option (see #3229
comments).

Was thinking of having a look at MMTF as well, but put it off when I saw it
wasn’t just another dir to enable in my rcsb mirror script - and going on
holiday rest of this week.

**Comment by JoaoRodrigues** [comment]
>Seems like the few mmCIF problems I looked at that had PDB equivalents had
the same issues, just mmCIF did not have the permissive option (see #3229
comments).

Indeed, same here. Which is good news. We might add a PERMISSIVE option to MMCIFParser just to standardize the interface of the parsers, what do you think?

>Was thinking of having a look at MMTF as well, but put it off when I saw it
wasn’t just another dir to enable in my rcsb mirror script - and going on
holiday rest of this week.

I have personally never loaded/opened/written a MMTF file so go ahead whenever you have time :)

**Comment by jgreener64** [comment]
> Maybe we can market Bio.PDB as a PDB validator

I did suggest this to them but never heard back. Perhaps we should mention the possibility at the bottom of every suggested change. I have also noticed anecdotally that the problematic structures I have found were all submitted via PDBe. I wonder if the different PDBs do different validation checks. It could be chance though.

**Comment by JoaoRodrigues** [comment]
Reported the first group of structures to RCSB: some 20 entries that our PDB and mmCIF parsers don't represent similarly due to mislabelings.

One issue that stands out is the handling of alt locations that are grouped by residue, as in, several different residues all with disordered atoms that are grouped res1Ares2Ares3A....res1Bres2Bres3B... I asked if we should handle such a case, given that the format spec is ambiguous for this case. If it turns out to be the case, we'll have to overhaul the parsers a bit not to expect residues/chains in sequential order.

**Comment by rob-miller** [comment]
> Indeed, same here. Which is good news. We might add a PERMISSIVE option to MMCIFParser just to standardize the interface of the parsers, what do you think?

Yes I think so.  In PDBParser we catch StructureBuilder exceptions and just warn, so the same approach could be used in MMCIFParser (and again if I get to play with MMTF :-)).

1ejg is an interesting case in point.  There's a Pro and two Ser coordinate sets defined at seqpos 22, but apparently the initial N doesn't move so it is assigned PRO with no altloc and there is no N for either of the SERs.  I think both the mmCIF and PDB files are wrong as they read the SER 22 residues with the N labeled as PRO, but we permissively allow the PDB file and break on the mmCIF file.
